In [191]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [192]:
# Load the given datasets
categorical = pd.read_csv('chronic_kidney_disease_categorical.csv')
numerical = pd.read_csv('chronic_kidney_disease_numerical.csv')

# Print the data
print(categorical)
print(numerical)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

The features and labels are different levels and measurements in the human body to predict kidney disease. The features are both categorical and numerical. Specific gravity, albumin, and sugar are three categorical features and age, blood pressure, and sodium levels are numerical features. 

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [193]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num_duplicated = numerical.drop_duplicates(subset=['unique_id'])
if (num_duplicated.shape[0] < numerical.shape[0]):
    print(numerical.shape[0] - num_duplicated.shape[0])
# Drop these duplicate rows


# Repeat the same for categorical dataset. Print the duplicate rows and drop them
cat_duplicated = categorical.drop_duplicates(subset=['unique_id'])
if (cat_duplicated.shape[0] < categorical.shape[0] ):
    print(categorical.shape[0] - cat_duplicated.shape[0])


30
30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

In [194]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined = pd.merge(cat_duplicated, num_duplicated, on='unique_id')
#Print the combined dataset
print(combined)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
385     789906  0.0  0.0    normal    normal  notpresent  notpresent   no   
386     490253  0.0  0.0    normal    normal  notpresent  notpresent   no   
387     492993  3.0  0.0       NaN    normal  notpresent  notpresent  yes   
388     273038  0.0  0.0    normal    normal  notpresent  notpresent   no   
389     583971  1.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm cad  ...    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc 

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [195]:
# Calculate the percentage of rows that contain atleast one missing value
has_missing = pd.isna(combined)
missing = 0
for seq in has_missing :
    if any(seq):
        missing += 1
percent = missing/(len(has_missing))*100
# Print %
print(percent)
# Drop these rows from the dataset
remove_missing = combined.dropna()
# Print the Dataset
print(remove_missing)

6.41025641025641
     unique_id   al   su       rbc      pc         pcc          ba  htn   dm  \
4       881763  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
5       114717  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
7       901528  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
8       433613  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
9       621332  1.0  0.0  abnormal  normal  notpresent  notpresent   no  yes   
..         ...  ...  ...       ...     ...         ...         ...  ...  ...   
382     996076  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
384     397388  2.0  2.0    normal  normal  notpresent     present  yes   no   
385     789906  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
386     490253  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
388     273038  0.0  0.0    normal  normal  notpresent  notpresent   no   no   

     cad  ...     bp  

## Part 3.4: Sort the dataset according to the Labels

In [196]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
combined_sorted = remove_missing.sort_values(by=['Target'])
combined_sorted = combined_sorted.reset_index(drop=True)
# Print the dataset
print(combined_sorted)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       534245  2.0  0.0    normal  abnormal     present  notpresent   no   
1       349892  1.0  3.0  abnormal  abnormal  notpresent  notpresent  yes   
2       343710  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   
4       514721  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
..         ...  ...  ...       ...       ...         ...         ...  ...   
148     489160  0.0  0.0    normal    normal  notpresent  notpresent   no   
149     530138  0.0  0.0    normal    normal  notpresent  notpresent   no   
150     261137  0.0  0.0    normal    normal  notpresent  notpresent   no   
151     322882  0.0  0.0    normal    normal  notpresent  notpresent   no   
152     273038  0.0  0.0    normal    normal  notpresent  notpresent   no   

      dm  cad  ...    bp    bgr     bu    sc    sod  pot  hemo   pcv     wb

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [197]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
cat_dummies = pd.get_dummies(combined_sorted, categorical_columns)
# Print the dataset
print(cat_dummies)

     unique_id   al   su   age    bp    bgr     bu    sc    sod  pot  ...  \
0       534245  2.0  0.0  45.0  70.0  117.0   52.0   2.2  136.0  3.8  ...   
1       349892  1.0  3.0  59.0  70.0  424.0   55.0   1.7  138.0  4.5  ...   
2       343710  2.0  0.0  61.0  80.0  173.0  148.0   3.9  135.0  5.2  ...   
3       102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0  7.6  ...   
4       514721  2.0  0.0  56.0  90.0  129.0  107.0   6.7  131.0  4.8  ...   
..         ...  ...  ...   ...   ...    ...    ...   ...    ...  ...  ...   
148     489160  0.0  0.0  28.0  60.0   79.0   50.0   0.5  145.0  5.0  ...   
149     530138  0.0  0.0  23.0  80.0  111.0   34.0   1.1  145.0  4.0  ...   
150     261137  0.0  0.0  35.0  60.0  105.0   39.0   0.5  135.0  3.9  ...   
151     322882  0.0  0.0  33.0  80.0  128.0   38.0   0.6  135.0  3.9  ...   
152     273038  0.0  0.0  47.0  60.0  117.0   22.0   1.2  138.0  3.5  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_ye

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [198]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for col in numerical_columns:
    mean = cat_dummies[col].mean()
    std = cat_dummies[col].std()

    least = mean - 3*std
    most = mean + 3*std

    filtered_columns = cat_dummies[(cat_dummies[col] < most) & (cat_dummies[col] > least)]

filtered_columns = filtered_columns.reset_index(drop=True)
# Print the dataset
print(filtered_columns)

     unique_id   al   su   age    bp    bgr     bu    sc    sod  pot  ...  \
0       534245  2.0  0.0  45.0  70.0  117.0   52.0   2.2  136.0  3.8  ...   
1       349892  1.0  3.0  59.0  70.0  424.0   55.0   1.7  138.0  4.5  ...   
2       343710  2.0  0.0  61.0  80.0  173.0  148.0   3.9  135.0  5.2  ...   
3       102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0  7.6  ...   
4       514721  2.0  0.0  56.0  90.0  129.0  107.0   6.7  131.0  4.8  ...   
..         ...  ...  ...   ...   ...    ...    ...   ...    ...  ...  ...   
147     489160  0.0  0.0  28.0  60.0   79.0   50.0   0.5  145.0  5.0  ...   
148     530138  0.0  0.0  23.0  80.0  111.0   34.0   1.1  145.0  4.0  ...   
149     261137  0.0  0.0  35.0  60.0  105.0   39.0   0.5  135.0  3.9  ...   
150     322882  0.0  0.0  33.0  80.0  128.0   38.0   0.6  135.0  3.9  ...   
151     273038  0.0  0.0  47.0  60.0  117.0   22.0   1.2  138.0  3.5  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_ye

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [199]:
# Normalize the all Numerical Attributes in the dataset.
normalized = (filtered_columns[numerical_columns]-filtered_columns[numerical_columns].mean())/filtered_columns[numerical_columns].std()
normalized = filtered_columns.replace(normalized)
# Print the dataset
print(normalized)

     unique_id   al   su       age        bp         bgr          bu  \
0       534245  2.0  0.0 -2.201446 -1.278195   -0.567218   -0.044568   
1       349892  1.0  3.0  0.499175 -1.278195  424.000000   -0.044568   
2       343710  2.0  0.0 -0.336732  0.563375  173.000000   -0.382226   
3       102677  3.0  0.0  0.499175 -1.278195   -0.659036  186.000000   
4       514721  2.0  0.0 -0.465333 -0.357410   -0.046916   -0.740987   
..         ...  ...  ...       ...       ...         ...         ...   
147     489160  0.0  0.0  0.627776  0.563375    0.136720   -0.719884   
148     530138  0.0  0.0 -0.079530  0.563375   -0.016310   -0.361122   
149     261137  0.0  0.0  1.399382  0.563375   -0.307067    0.905096   
150     322882  0.0  0.0  1.206480  0.563375    0.044902    3.437532   
151     273038  0.0  0.0  0.627776  0.563375   -0.567218   -0.255604   

           sc       sod       pot  ...  cad_no  cad_yes  appet_good  \
0    2.200000  1.087021  3.800000  ...    True    False        T

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

the unique_id column is not needed because it is only an identifier and doesn't show any relationship or predictions across the dataset. For the dummy categories, there are 2 columns that display essentially the same data, so you can remove one of them. 

In [200]:
#Remove that column
final = normalized.drop(columns=['unique_id', 'rbc_abnormal', 'pc_abnormal', 'pcc_notpresent', 'ba_notpresent', 'htn_no', 'dm_no', 'cad_no', 'appet_poor', 'pe_no', 'ane_no', 'Target_notckd'])
# Print the dataset
print(final)

      al   su       age        bp         bgr          bu        sc       sod  \
0    2.0  0.0 -2.201446 -1.278195   -0.567218   -0.044568  2.200000  1.087021   
1    1.0  3.0  0.499175 -1.278195  424.000000   -0.044568  1.700000  1.483566   
2    2.0  0.0 -0.336732  0.563375  173.000000   -0.382226  3.900000  0.029567   
3    3.0  0.0  0.499175 -1.278195   -0.659036  186.000000  1.389586  0.029567   
4    2.0  0.0 -0.465333 -0.357410   -0.046916   -0.740987  6.700000  0.029567   
..   ...  ...       ...       ...         ...         ...       ...       ...   
147  0.0  0.0  0.627776  0.563375    0.136720   -0.719884  0.500000  1.087021   
148  0.0  0.0 -0.079530  0.563375   -0.016310   -0.361122  1.100000  1.087021   
149  0.0  0.0  1.399382  0.563375   -0.307067    0.905096  0.500000  0.029567   
150  0.0  0.0  1.206480  0.563375    0.044902    3.437532  0.600000  0.029567   
151  0.0  0.0  0.627776  0.563375   -0.567218   -0.255604  1.200000  1.483566   

          pot       hemo  .

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [201]:
# Export the dataframe to a new csv file
final.to_csv('results.csv')
# Export the dataframe to a new json file
final.to_json('results.json')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [202]:
df = pd.read_csv('results.csv')
print(df.head(15))

    Unnamed: 0   al   su       age        bp         bgr          bu  \
0            0  2.0  0.0 -2.201446 -1.278195   -0.567218   -0.044568   
1            1  1.0  3.0  0.499175 -1.278195  424.000000   -0.044568   
2            2  2.0  0.0 -0.336732  0.563375  173.000000   -0.382226   
3            3  3.0  0.0  0.499175 -1.278195   -0.659036  186.000000   
4            4  2.0  0.0 -0.465333 -0.357410   -0.046916   -0.740987   
5            5  3.0  4.0  0.692076 -1.278195  253.000000   -0.297811   
6            6  3.0  4.0 -2.394347 -1.278195  264.000000   -0.213397   
7            7  3.0  0.0  0.692076 -1.278195   -0.291764   -0.276707   
8            8  2.0  1.0 -2.072845 -0.357410  273.000000  235.000000   
9            9  4.0  0.0  1.270781  0.563375   -0.123431   -0.762091   
10          10  3.0  1.0 -2.072845  0.563375  214.000000   -0.403329   
11          11  3.0  0.0  0.113372 -1.278195   -0.521309   -0.340018   
12          12  1.0  0.0  0.563475  0.563375  163.000000   -0.40

### Paste the markdown table here

| Unnamed: 0 |   al |   su |        age |        bp |          bgr |          bu |     sc |       sod |      pot | ... | pc_normal | pcc_present | ba_present | htn_yes | dm_yes | cad_yes | appet_good | pe_yes | ane_yes | Target_ckd |
|-------------|------|------|------------|------------|----------------|----------------|--------|-----------|----------|-----|-----------|-------------|------------|---------|--------|---------|-------------|--------|---------|------------|
|           0 |  2.0 |  0.0 | -2.201446 | -1.278195 |     -0.567218 |     -0.044568 |  2.2000 |  1.087021 |  3.800000 | ... |     False |        True |      False |    False |   False |   False |       True |  False |   False |       True |
|           1 |  1.0 |  3.0 |  0.499175 | -1.278195 |    424.000000 |     -0.044568 |  1.7000 |  1.483566 |  4.500000 | ... |     False |       False |      False |     True |    True |    True |       True |  False |   False |       True |
|           2 |  2.0 |  0.0 | -0.336732 |  0.563375 |    173.000000 |     -0.382226 |  3.9000 |  0.029567 |  5.200000 | ... |     False |       False |      False |     True |    True |    True |      False |   True |    True |       True |
|           3 |  3.0 |  0.0 |  0.499175 | -1.278195 |     -0.659036 |    186.000000 |  1.3896 |  0.029567 |  7.600000 | ... |     False |       False |      False |     True |    True |    True |      False |   True |    True |       True |
|           4 |  2.0 |  0.0 | -0.465333 | -0.357410 |     -0.046916 |     -0.740987 |  6.7000 |  0.029567 |  4.800000 | ... |     False |       False |      False |     True |    False |   False |       True |  False |   False |       True |
|           5 |  3.0 |  4.0 |  0.692076 | -1.278195 |    253.000000 |     -0.297811 | 11.9000 |  0.426112 |  5.600000 | ... |      True |       False |      False |     True |    True |   False |      False |   True |   False |       True |
|           6 |  3.0 |  4.0 | -2.394347 | -1.278195 |    264.000000 |     -0.213397 |  2.7000 |  1.087021 |  0.044587 | ... |     False |       False |      False |     True |    True |    True |       True |   True |   False |       True |
|           7 |  3.0 |  0.0 |  0.692076 | -1.278195 |     -0.291764 |     -0.276707 |  1.7000 |  1.087021 |  4.700000 | ... |     False |       False |      False |     True |    True |   False |       True |  False |   False |       True |
|           8 |  2.0 |  1.0 | -2.072845 | -0.357410 |    273.000000 |   235.000000 | 14.2000 |  0.426112 |  3.400000 | ... |     False |       False |      False |     True |    True |    True |      False |   True |    True |       True |
|           9 |  4.0 |  0.0 |  1.270781 |  0.563375 |     -0.123431 |     -0.762091 | -0.1453 |  0.029567 |  4.900000 | ... |     False |       False |       True |    False |   False |   False |      False |  False |   False |       True |
|          10 |  3.0 |  1.0 | -2.072845 |  0.563375 |    214.000000 |     -0.403329 |  3.9000 |  1.087021 |  4.900000 | ... |     False |        True |       True |     True |    True |   False |       True |   True |   False |       True |
|          11 |  3.0 |  0.0 |  0.113372 | -1.278195 |     -0.521309 |     -0.340018 |  2.6000 | -0.234796 |  5.700000 | ... |      True |       False |      False |     True |   False |    True |      False |  False |   False |       True |
|          12 |  1.0 |  0.0 |  0.563475 |  0.563375 |    163.000000 |     -0.403329 |  3.3000 | -0.102615 |  0.044587 | ... |      True |       False |      False |     True |   False |    True |       True |  False |   False |       True |
|          13 |  4.0 |  2.0 | -1.236938 | -1.278195 |    163.000000 |     -0.297811 |  7.2000 |  0.426112 |  4.600000 | ... |     False |       False |       True |     True |    True |   False |       True |   True |   False |       True |
|          14 |  4.0 |  0.0 |  1.077879 |  0.563375 |     0.136720 |   163.000000 |  7.7000 |  1.087021 |  3.800000 | ... |     False |       False |      False |     True |   False |    True |       True |  False |   True  |       True |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [203]:
# Code to rename all the columns in the dataset
import pandas as pd

# Create a mapping for renaming the columns
column_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class',
    'target_ckd': 'target_ckd'
}

# Rename the columns of the DataFrame
df.rename(columns=column_mapping, inplace=True)



## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [204]:

# Now, we perform the equivalent of the SQL query
result = df.groupby('Target_ckd').size().reset_index(name='count')

# Optionally, sort the result by count in descending order
result = result.sort_values(by='count', ascending=False)

# Display the resulting DataFrame
print(result)


   Target_ckd  count
0       False    111
1        True     41


*Describe what the above code does here*

The code counts how many rows are targets for chronic kidney disease and how many are not and groups them together with the counts.